### DISCUSSION TITLE
Jetson Nano flashing - image not aligned

### DISCUSSION THREAD

**Discussion by tirankas** [OE4T/tegra-demo-distro#98] [General]
Hello,
I was able to build the correct images for NX Xavier with mender support and everything (well almost..) works fine.

Trying to build the same for Nano (production som), when I try to flash the tegraflash, I see the following error in the console (during flashing the .ext4 rootfs):
Written less bytes than expectedHALT: spinning forever...

Is there any setup I need to make in order to make the ext4 image block aligned?

Thanks

**Comment by madisongh** [comment]
> Written less bytes than expectedHALT: spinning forever...

I think I've seen this when the filesystem to be written is larger than the partition size.  Can you provide more information on how you came to the conclusion that this has something to do with block alignment?

**Reply by tirankas** [reply]
When I switched to squashfs rootFs I got an error message during flash of the rootfs complaining on image is not multiple of block size (1048576)

**Reply by tirankas** [reply]
I initially this has something to do with my generated squashfs image so I went back to the original settings of ext4 which show the error at 100% of flashing meaning probably block alignment...
I use the same sources to build for Xavier with no issues.
Is there a difference in block sizes between Xavier and nano?

**Reply by madisongh** [reply]
It's just a difference in the flashing tools for Nano vs TX2/Xavier, based on what I see in the L4T flashing scripts.  Have you tried padding out the squashfs image to be an even multiple of 1MiB in size before flashing?

**Comment by tirankas** [comment]
Yes, I did, not booting up after flashing in this case.
I will check maybe we made a mistake mixing the flashing tools between these two machine...

**Comment by tirankas** [comment]
Hi Matt,
I see we have removed this from the machine configuration file:
MACHINEOVERRIDES = "jetson-nano-devkit-emmc:cot:${MACHINE}"

Can you please explain what is this for?

Thanks

**Reply by madisongh** [reply]
See [this wiki page](https://github.com/OE4T/tegra-demo-distro/wiki/Tegra-&-Uboot-Flattened-Image-Tree-(FIT-image)).

**Reply by tirankas** [reply]
Looking at the wiki, I see the fit image support is not available for the tegrademo-mender.
I am using the 32.5.0 branch, is this still true, I mean not supported for the mender distro?

I am building the nano for the mender distro, this is where I see the alignment issues coming from...
I checked the .ext4 in the nano and xavier, both are not 1M aligned, but the later works fine...

Thanks

**Reply by ichergui** [reply]
Hey @tirankas 
That's true, FIT image support is availble when using ditro `tegrademo-mender`.
Do you need it ? if so I can make it available shortly.
As I mentioned in the [Wiki page](https://github.com/OE4T/tegra-demo-distro/wiki/Tegra-&-Uboot-Flattened-Image-Tree-(FIT-image)#build). FIT-image is supported only for `tegrademo` distro at the moment.

**Reply by tirankas** [reply]
Hey @ichergui 
I kind of not seen this wiki before and was not aware of the needed adaptation for the mender.
Currently my nano machine is with the fit image and was built for mender (since mender is a hard requirment for both our Nano and Xavier). This is where I see the issues not being able to flash the initial image (due to image not aligned). I am not sure this is my issue but worth the try.
Also, I noticed that both rootfs from Xavier and Nano are not 1M aligned, but the flashing tool complains only for the Nano...

If you have the mender distro changes, would appreciate if you provide it.

Thanks

**Reply by ichergui** [reply]
Hey @tirankas 
The [wiki page](https://github.com/OE4T/tegra-demo-distro/wiki/Tegra-&-Uboot-Flattened-Image-Tree-(FIT-image)) was shared by @madisongh 
Please see [here](https://github.com/OE4T/tegra-demo-distro/discussions/98#discussioncomment-738192)
Could you please share the steps that you used ti build the image ? I'm asking because  `dunfell-l4t-r32.5.0` brach is working fine to me with different jetson platform.

**Reply by tirankas** [reply]
Hi @ichergui 
I will share the steps after I make a clean demo build (remove my layer and use the tegrademo clean layer).
Do youean that the 32.5.0 branch works for the nano with the mender distro? No need for distro changes like what's written in the wiki for 32.4.3?

Thanks

**Comment by tirankas** [comment]
Hi @ichergui 
I checked with the clean tegra demo distro build and facing the same issue (with the ext4 original FS):

**Reply by tirankas** [reply]
U-Boot 2020.04 (Apr 20 2021 - 13:47:00 +0000)

SoC: tegra210
Model: NVIDIA Jetson Nano Developer Kit
Board: NVIDIA P3450-0002
DRAM:  4 GiB
MMC:   sdhci@700b0000: 1, sdhci@700b0600: 0
Loading Environment from MMC... *** Warning - bad CRC, using default environment

In:    serial
Out:   serial
Err:   serial
Net:   No ethernet found.
Hit any key to stop autoboot:  0
Saving Environment to MMC... Writing to redundant MMC(0)... OK
Retrieving file: /boot/extlinux/extlinux.conf
Error reading config file

Following is the build info:
. ./setup=env --machine jetson-nano-devkit-emmc-cot --distro tegrademo-mender build-nano-devkit
bitbake demo-image-full

I am on dunfell-l4t-r32.5.0

Thanks

**Reply by ichergui** [reply]
Hey @tirankas 
This is an expected behavior, as I mentioned yesterday `FIT-image` support is available only when using `tegrademo`  distro.

Please try the following builds steps:

```
$ . ./setup-env --machine jetson-nano-emmc-cot --distro tegrademo build-nano-devkit
$ bitbake demo-image-full
```

I will work on it.
I will keep you updated concerning `FIT-image` with `mender` support

**Reply by tirankas** [reply]
Hi @ichergui 
At least now I understand where my issue comes from :-)
I changed the tegrademo to squashfs and I get the same issue I got with my machine: "image is not multiple of block size (1048576)"

Do you have estimate when the mender distro is available for the nano?
Can I help somehow to make this faster?

Thanks

**Comment by ichergui** [comment]
Hey @tirankas 

PR #102  is ready for review, it is for the `master` branch, I back port the commit to other branches once the PR is merged. 
The tests were done `with` and `without` **signing FIT image** on `Jetson nano devkit (SDcard version)`
A/B system update is working fine to me.
Enjoy

```
U-Boot 2021.01 (Feb 28 2021 - 13:04:45 +0000)

SoC: tegra210
Model: NVIDIA Jetson Nano Developer Kit
Board: NVIDIA P3450-0000
DRAM:  4 GiB
MMC:   sdhci@700b0000: 1, sdhci@700b0600: 0
Loading Environment from SPIFlash... SF: Detected mx25u3235f with page size 256 Bytes, erase size 4 KiB, total 4 MiB
OK
In:    serial
Out:   serial
Err:   serial
Net:   No ethernet found.
Hit any key to stop autoboot:  0 
39309158 bytes read in 1697 ms (22.1 MiB/s)
## Loading kernel from FIT Image at 90000000 ...
   Using 'conf-1' configuration
   Verifying Hash Integrity ... sha256,rsa2048:dev+ OK
   Trying 'kernel-1' kernel subimage
     Description:  Linux kernel
     Type:         Kernel Image
     Compression:  uncompressed
     Data Start:   0x90000124
     Data Size:    39307272 Bytes = 37.5 MiB
     Architecture: AArch64
     OS:           Linux
     Load Address: 0x80080000
     Entry Point:  0x80080000
     Hash algo:    sha256
     Hash value:   21829f4c5b5b0c9b3730da8ca7d69f31def901c53d4b0995ae804088d2548612
   Verifying Hash Integrity ... sha256+ OK
## Flattened Device Tree blob at 83100000
   Booting using the fdt blob at 0x83100000
   Loading Kernel Image
ERROR: reserving fdt memory region failed (addr=0 size=0)
ERROR: reserving fdt memory region failed (addr=0 size=0)
   Using Device Tree in place at 0000000083100000, end 000000008317e6d9
copying carveout for /host1x@50000000/dc@54200000...
copying carveout for /host1x@50000000/dc@54240000...

Starting kernel ...

[    0.000000] Booting Linux on physical CPU 0x0
[    0.000000] Linux version 4.9.201-l4t-r32.5+ga3633e2e1cf4 (oe-user@oe-host) (gcc version 8.3.0 (GCC) ) #1 SMP PREEMPT Thu May 6 13:07:24 UTC 2021

```

```
root@jetson-nano-devkit-cot:~# mender show-artifact
demo-image-base_3.4-0fc43c3-dirty-ichergui
root@jetson-nano-devkit-cot:~# 
```

**Comment by tirankas** [comment]
Thanks @ichergui !
I will give it a try and let you know

**Comment by tirankas** [comment]
Hi @ichergui 
This is what I get:

** Unrecognized filesystem type **
Wrong Image Format for bootm command
ERROR: can't get kernel image!

Thanks

**Reply by ichergui** [reply]
Hmm, 
Did you flash your device with script `doflash.sh` ? because we need to update `SPI` flash. 
if no, please flash your device via `doflash.sh`
Are you using that you are using `master` branch with my changes ?
Are you using my fork of  [tegra-demo-distro](https://github.com/ichergui/tegra-demo-distro) ?
Please share the steps that you used for the build ?

**Reply by madisongh** [reply]
@tirankas Is this with your squashfs rootfs?  If so, did you configure U-Boot with squashfs support?

**Reply by tirankas** [reply]
@madisongh Does U-Boot supports the squashfs? I looked at the u-boot-tegra and didnt see any squashfs FS support...

**Reply by tirankas** [reply]
Sorry, I found it on the master branch, will configure appropriately...

**Reply by madisongh** [reply]
I think it's a relatively recent addition to U-Boot, so you may not find it in the older version on the dunfell-based branches.

**Comment by tirankas** [comment]
Hi Matt,
I am checking, maybe I changed back to squashfs by mistake, the error itself looks like related to this. Will confirm.
Hi @ichergui,
I am not using your fork, I simply took your commit as is and applied to my build via a bbappend file.
I noticed that in the master branch there are more fit related commits that I don't pick up, like in the recipes-bsp (fit-image-.. bb) and in the machine too.
I will try to cherry pick those changes too.

Will update.

Thanks

**Comment by tirankas** [comment]
Hi @ichergui 
What is the tegra-fitimage_1.0.bb in your fork (under recipes-bsp)?
DO I need this too?

Thanks

**Reply by ichergui** [reply]
Hey @tirankas 
`tegra-fitimage` recipe is needed because it used for FIT image packaging and installation into the `rootfs`
Yes, you need it.
BTW, the PR was merged.

**Comment by tirankas** [comment]
When I pick up the tegra-fitimage, I get this error:
file /boot/fitImage conflicts between attempted installs of kernel-image-fitimage-4.9.201-l4t-r32.5+gc68230c5b46a-4.9.201+git0+c68230c5b4-r0.jetson_nano_emmc_cot and tegra-fitimage-1.0-r0.jetson_nano_emmc_cot

I will do a clean of these two packages and will check.

**Reply by ichergui** [reply]
Are you using the `master` branch ?
`tegra-fitimage` recipe is needed only when using `master` branch

**Comment by tirankas** [comment]
I am not in the master branch, this is why I asked if I should pick this.
I am in the 32.5.0, should I remove this recipe?
Will mender distro work fine with picking your commit without this additional bb file?

**Reply by ichergui** [reply]
Please remove it.
I'm working on it, I will push a PR soon for enabling the `FIT image` support with mender with branch `dunfell-l4t-r32.5.0`
I have to complete my tests by using different machines.

**Comment by tirankas** [comment]
Thanks Guys for your help!
I am able to add the squashfs support to uboot, it works fine now.
There still seems to be a problem with the mender, dont think this comes from the fact that I run the root as squashfs, maybe you have seen such error:

mender-client-resize-data-part[3749]: cat: can't open '/sys/block/mmcblk0/mmcblk0p23/size': No such file or directory

**Reply by tirankas** [reply]
Probably something wrong with my partition table, checking...

**Reply by ichergui** [reply]
This issue was fixed in the `master` branch. 
Please take a look [here](https://github.com/OE4T/tegra-demo-distro/commit/885933ecffcf750eea89a0ff5119cd69875c7bdd)

**Reply by ichergui** [reply]
As I told you I will push new PR soon concerning `dunfell-l4t-r32.5.0`

**Reply by ichergui** [reply]
* Target finish booting
```
root@jetson-nano-devkit-emmc-cot:~# hostnamectl 
   Static hostname: jetson-nano-devkit-emmc-cot
         Icon name: computer
        Machine ID: b7b2510c894e4e7eba3518bd29f3dc82
           Boot ID: dda225c47da84d85a2805fc037195816
  Operating System: OE4Tegra Demonstration Distro with Mender 3.1-08f27cc-dirty-ichergui
            Kernel: Linux 4.9.201-l4t-r32.5+gc68230c5b46a
      Architecture: arm64
root@jetson-nano-devkit-emmc-cot:~# mender show-artifact
demo-image-base_3.1-08f27cc-dirty-ichergui
root@jetson-nano-devkit-emmc-cot:~# 
```
* mender install 
```
root@jetson-nano-devkit-emmc-cot:~# mender install /tmp/ilies/demo-image-base-jetson-nano-devkit-emmc-cot-20210520164640.mender
INFO[0000] Loaded configuration file: /var/lib/mender/mender.conf 
INFO[0000] Loaded configuration file: /etc/mender/mender.conf 
INFO[0000] Mender running on partition: /dev/mmcblk0p1  
INFO[0000] Start updating from local image file: [/tmp/ilies/demo-image-base-jetson-nano-devkit-emmc-cot-20210520164640.mender] 
Installing Artifact of size 202966016...
INFO[0000] No public key was provided for authenticating the artifact 
INFO[0000] Update Module path "/usr/share/mender/modules/v3" could not be opened (open /usr/share/mender/modules/v3: no such file or directory). Update modules will not be available 
INFO[0000] Opening device "/dev/mmcblk0p18" for writing 
INFO[0000] Native sector size of block device /dev/mmcblk0p18 is 512 bytes. Mender will write in chunks of 1048576 bytes 
.............................................................. - 100 %
INFO[0032] All bytes were successfully written to the new partition 
INFO[0032] The optimized block-device writer wrote a total of 4743 frames, where 0 frames did need to be rewritten (i.e., skipped) 
INFO[0033] Wrote 4972347392/4972347392 bytes to the inactive partition 
INFO[0033] Enabling partition with new image installed to be a boot candidate: 18 
Use -commit to update, or -rollback to roll back the update.
At least one payload requested a reboot of the device it updated.
root@jetson-nano-devkit-emmc-cot:~# 

```
* mender commit 
```
root@jetson-nano-devkit-emmc-cot:~# mender commit 
INFO[0000] Loaded configuration file: /var/lib/mender/mender.conf 
INFO[0000] Loaded configuration file: /etc/mender/mender.conf 
INFO[0000] Mender running on partition: /dev/mmcblk0p18 
INFO[0000] Update Module path "/usr/share/mender/modules/v3" could not be opened (open /usr/share/mender/modules/v3: no such file or directory). Update modules will not be available 
Committing Artifact...
INFO[0000] Committing update                            
root@jetson-nano-devkit-emmc-cot:~# 
```

**Reply by tirankas** [reply]
I will wait for you PR, the changed suggested above still gives me the same error but probably was not rebuild correctly.
I will pick the PR once done

**Reply by ichergui** [reply]
Hey @tirankas 
The PR was merged. refer to [tegra-demo-distro](https://github.com/OE4T/tegra-demo-distro/tree/dunfell-l4t-r32.5.0)
Please let me know if you need help

**Reply by tirankas** [reply]
Thanks @ichergui 
Appreciate your help, it's ok for me now...
